File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        #maintain a dictionary for braceket pairs
        pairs = {'(':')','{':'}','[':']'}
        stack = []
        
        for bracket in s:
            if bracket in pairs:
                stack.append(bracket)
            else:
                if not stack:
                    return False
                openingBracket = stack.pop()
                if pairs[openingBracket] != bracket:
                    return False
        return not stack
        """
        :type s: str
        :rtype: bool
        """
        

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        //maintain hashMap for bracket pairs
        Map<Character, Character> pairs = new HashMap<>();
        pairs.put('(',')');
        pairs.put('{','}');
        pairs.put('[',']');
        
        //maintian stack for opening brackets
        Stack<Character> stack = new Stack<>();
        char brackets[] = s.toCharArray();
        for(char bracket : brackets){
            if(pairs.containsKey(bracket))
                stack.push(bracket);
            else{
                if(stack.isEmpty())
                    return false;
                char openingBracket = stack.pop();
                char closingBracket = pairs.get(openingBracket);
                if(closingBracket != bracket)
                    return false;
            }
        }
        return stack.isEmpty();
    }
}